Add image cache for image store

kikimo opened the pull request:
In a system with hundreds of docker images stored, periodically calling of
api `/images/json` from process like kubelet will make dockerd daemon
a cpu hog. This patch add cache for docker images to avoid computing
sha256 hash and json decoding each time they are retrieved thus reducing
dockerd cpu overhead drastically.

Signed-off-by: Wenlin Wu <fzuwwl@gmail.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Improve dockerd's performance on systems with hundreds of docker images stored.

**- How I did it**
Add image cache for image store.

**- How to verify it**
To verify, you need to prepare a system with hundreds of docker images stored, and a simple shell script to query docker api `/images/json` every seconds is enough to create the payload. You will see the cpu usage of dockerd decrease drastically after applying this patch.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Add image cache for image store.

**- A picture of a cute animal (not mandatory but encouraged)**

AkihiroSuda commented:
Benchmark result please?

AkihiroSuda commented on image/store.go:
```diff
@@ -201,6 +203,26 @@ func (is *store) Search(term string) (ID, error) {
+	return img, nil
+}
+
+func (is *store) doGet(id ID) (*Image, error) {
```
Please add godoc about `doGet` vs `Get`

AkihiroSuda commented on image/store.go:
```diff
@@ -201,6 +203,26 @@ func (is *store) Search(term string) (ID, error) {
+	}
+
+	is.Lock()
+	defer is.Unlock()
```
nit: you can easily unlock without defer

AkihiroSuda commented on image/store.go:
```diff
@@ -247,7 +269,9 @@ func (is *store) Delete(id ID) ([]layer.Metadata, error) {
 	}
+
 	delete(is.images, id)
+	delete(is.imageCache, id)
```
In addition to this deletion on `Delete()`, the cache should be also periodically deleted?

AkihiroSuda replied:
We might be able to use https://github.com/hashicorp/golang-lru

kikimo replied:
I've been considering using this lib days before, but I'm not sure whether it worth doing considering the complexity we introduce. Anyway, there are still bug in the current implementation, I'm still working on it and I'll do more test later.

kikimo replied:
@AkihiroSuda I just replace map with fixed size golang-lru cache, the cache size if set to 1024.

kikimo commented:
> Benchmark result please?

# Benchmark for Docker with Image Cache

## 1. A server with 337 docker images:

```bash
# docker images | wc -l
337
```

## 2. Shell script image_query.sh to create payload:

```bash
#!/bin/bash

while true; do
    curl --unix-socket /var/run/docker.sock http://./images/json >/dev/null
    sleep 1
done
```

##  3. Test result on unpatched dockerd(master branch e9b4655bc9) 

```bash
[root@k8s-node bin]# ps aux | grep dockerd
root      52963 53.9  0.1 6515372 133424 ?      Ssl  00:03   0:32 /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock --metrics-addr=0.0.0.0:9323 --experimental --data-root=/data/docker/ --ip-masq=false --log-opt max-file=5 --log-opt max-size=100m
root      53370  0.0  0.0 112720  2184 pts/0    S+   00:04   0:00 grep --color=auto dockerd

[root@k8s-node bin]# pidstat  -u -p 52963 1
Linux XXXXXXXX (k8s-node) 	08/18/2020 	_x86_64_	(48 CPU)

12:04:13 AM   UID       PID    %usr %system  %guest    %CPU   CPU  Command
12:04:14 AM     0     52963   34.65    0.99    0.00   35.64    14  dockerd
12:04:15 AM     0     52963   19.00    0.00    0.00   19.00    14  dockerd
12:04:16 AM     0     52963   53.00    3.00    0.00   56.00    14  dockerd
12:04:17 AM     0     52963   80.00    4.00    0.00   84.00    14  dockerd
12:04:18 AM     0     52963   61.00    1.00    0.00   62.00    14  dockerd
12:04:19 AM     0     52963   53.00    3.00    0.00   56.00    14  dockerd
12:04:20 AM     0     52963   19.00    0.00    0.00   19.00    14  dockerd
12:04:21 AM     0     52963   47.00    1.00    0.00   48.00    14  dockerd
12:04:22 AM     0     52963   62.00    2.00    0.00   64.00    14  dockerd
12:04:23 AM     0     52963   55.00    2.00    0.00   57.00    14  dockerd
12:04:24 AM     0     52963   60.00    1.00    0.00   61.00    14  dockerd
```

We can see that dockerd sonsumed about 60% of a cpu most of the time.

## 4. Test result on dockerd with image cache from master branch

```bash
[root@k8s-node bin]# ps aux | grep dockerd
root      54419 56.0  0.0 6151832 65400 ?       Ssl  00:08   0:02 /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock --metrics-addr=0.0.0.0:9323 --experimental --data-root=/k8sdata/docker/ --ip-masq=false --log-opt max-file=5 --log-opt max-size=100m
root      54631  0.0  0.0 112720  2244 pts/0    S+   00:08   0:00 grep --color=auto dockerd

[root@k8s-node-gpu-cicd-d-033030 bin]# pidstat  -u -p 54419 1
Linux XXXXXXXX (k8s-node) 	08/18/2020 	_x86_64_	(48 CPU)

12:09:03 AM   UID       PID    %usr %system  %guest    %CPU   CPU  Command
12:09:04 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:05 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:06 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:07 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:08 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:09 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:10 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:11 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:12 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:13 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:14 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:15 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:16 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:17 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:18 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:19 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
12:09:20 AM     0     54419    1.00    0.00    0.00    1.00     7  dockerd
12:09:21 AM     0     54419    0.00    0.00    0.00    0.00     7  dockerd
```

we can see that dockerd with image cache patch perform much better in a system with 337 docker images.

thaJeztah commented:
Discussing with @cpuguy83 @tonistiigi, and we're a bit concerned about the memory overhead from this; that, and if we reach the 1024 limit, effectively caching would not do anything at all.

As an alternative, we could have a look at having a special function for `docker images`  / `GET /images/json` that has less overhead and only does the bare minimal needed to produce output for that endpoint.

kikimo commented:
> Discussing with @cpuguy83 @tonistiigi, and we're a bit concerned about the memory overhead from this; that, and if we reach the 1024 limit, effectively caching would not do anything at all.
> 
> As an alternative, we could have a look at having a special function for `docker images` / `GET /images/json` that has less overhead and only does the bare minimal needed to produce output for that endpoint.

@thaJeztah @cpuguy83 @tonistiigi What about we expand `type imageMeta struct` in `images/store.go` to include  `ParentID`,   `Created`, `Labels` fields? This way could not only save the time to retrieve and decoding data from disk repeatly but also keep  the memory overhead in an acceptable level, most import we could still fullfill the computation of  `docker images` / `GET /images/json`.

kikimo committed:
Add image cache for image store

In a system with hundreds of docker images stored, periodically calling of
api `/images/json` from process like kubelet will make dockerd daemon
a cpu hog. This patch add `image.Meta` that stores `Image` info like `ChainID`
and `Layer` info thus avoid retrieving and decoding image info from disk repeatly.

Signed-off-by: Wenlin Wu <fzuwwl@gmail.com>

chunhualiu commented:
docker 20.10.10, still has this issue.
Why this fix was not merged?
